File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-3.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        maxLength = 0
        if len(s) <= 0: return maxLength

        check = set()
        right = 0
        left = 0
        while right < len(s):
            if s[right] not in  check:
                check.add(s[right])
            elif s[right] in check:
                maxLength = max(maxLength, len(check))
                right = left
                left += 1
                check.clear()
            right += 1

        
        return max(maxLength, len(check))

File: Data Structures & Algorithms/sliding-window-maximum/submission-3.java
class Solution {
    private int findMax(int[] array, int start, int end){
        int maxValue = array[start];
        for(int i = start; i < end; i++){
            maxValue = Math.max(maxValue, array[i]);
        }
        return maxValue;
    }

    public int[] maxSlidingWindow(int[] nums, int k) {
        List<Integer> list = new ArrayList<>();

        int left = 0;
        int right = k;

        int currentMax = findMax(nums, left, right);

        while(right < nums.length){
            list.add(currentMax);
            if(nums[left++] == currentMax) currentMax = findMax(nums, left, right);
            if(nums[right] > currentMax) currentMax = nums[right];
            right++;
        }

        list.add(currentMax);

        return list.stream().mapToInt(Integer::intValue).toArray();
    }
}
